Flank/flank
# feat: 1665 implement custom sharding -- android

Android part for #1665

### [Detailed documentation](https://github.com/Flank/flank/blob/1665-implement-custom-sharding-android/docs/feature/1665-custom-sharding.md)

This is the first implementation of the custom sharding feature, below assumptions were made:
1. flank does not validate provided json -- we trust the user that he/she knows what is doing
2. `sharding-json` is the source of truth -- no smart sharding applied
3. custom sharding can be launched together with robo/game loop test

## Test Plan
> How do we know the code works?

1. build flank `./gradlew flankFullRun`
2. create config yml 
    ```
    gcloud:
      app: [PATH]/app-debug.apk
      robo-script: [PATH]/MainActivity_robo_script.json
      use-orchestrator: false
    flank:
      max-test-shards: 2
      additional-app-test-apks:
        - test: [PATH]/app-single-success-debug-androidTest.apk
        - test: [PATH]/app-multiple-error-debug-androidTest.apk
        - test: [PATH]/app-multiple-success-debug-androidTest.apk
      disable-results-upload: true
      disable-usage-statistics: true
    ```
3. run `flank firebase test android run -c=[YOUR CONFIG] --dump-shards`
4. copy `android_shards.json`, rename it and play around with shards, example: add another shard, move tests etc
5. add `sharding-json: [PATH TO YOUR CUSTOM SHARDING JSON]` to your config yaml
6. run `flank firebase test android run -c=[YOUR CONFIG] --dump-shards`
7. observe new sharding in `android_shards.json` (notice that shards are properly assigned to app-test apk pairs)
8. run `flank firebase test android run -c=[YOUR CONFIG]`
9. test run should proceed and end normally with sharding that you created

#### Note:
* matrices ids don't matter when preparing custom sharding
* shards id as well

## Checklist
